File: 416-partition-equal-subset-sum/partition-equal-subset-sum.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) 
    {
        int sum = accumulate(nums.begin(),nums.end(),0);
        int n = nums.size();

        if ( sum % 2 !=  0) return false;

        int target = sum / 2;
        vector<vector<int>> dp(n+1,vector<int>(target+1,false));

        for(int i= 0; i <= n ; i++) dp[i][0] = true;

        for(int i=1 ;i <= n ;i++)
        {
            for(int s=1 ; s <= target ;s++)
            {
                dp[i][s] = dp[i-1][s]; //Skipping current item

                if ( nums[i-1] <= s)
                {
                    dp[i][s] = dp[i-1][s] || dp[i-1][s - nums[i-1]];
                }
            }
        }

        return dp[n][target];
        
    }
};

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        sumCheck = sum(nums)

        if sumCheck%2 != 0:
            return False

        n = len(nums)
        reqSum = sumCheck // 2

        dp = [ [False for _ in range(reqSum+1)] for _ in range(n+1)]
        
        for i in range(0,n+1):
            dp[i][0] = True

        for i in range(1,n+1):
            for j in range(1,reqSum+1):

                if nums[i-1] <= j:
                    dp[i][j] = dp[i-1][j-nums[i-1]] or dp[i-1][j]
                else:
                    dp[i][j] = dp[i-1][j]

        return dp[n][reqSum]
        